app/controllers/pull_requests_controller.rb

```diff
@@ -5,52 +5,102 @@ class PullRequestsController < ApplicationController
   # Receives a webhook from both Github and Stash requesting a new build
   # TODO(rob): pretty good chance Github's request format has changed and this action won't work anymore. Verify.
   def build
-    if payload['host']
-      @repository = Repository.lookup(host: payload["host"], namespace: payload["repository"]["key"], name: payload["repository"]["slug"])
+    if params['payload']
+      # from stash
+      handle_stash_request(JSON.parse(params['payload']))
     else
-      url = payload['repository']['url'] || payload['repository']['ssh_url']
-      ssh_url = RemoteServer.for_url(url).canonical_repository_url
-      @repository = Repository.lookup_by_url(ssh_url)
+      # from github
+      handle_github_request(params)
+    end
+    render json: {message: "Thanks!"}
+  end
+
+  def handle_stash_request(payload)
+    repo_info = {
+      host: payload['host'],
+      namespace: payload['repository']['key'],
+      name: payload['repository']['slug'],
+      url: payload['repository']['url'] || payload['repository']['ssh_url']
+    }
+    @repo = get_repo(repo_info)
+
+    if payload['pull_request'] && active_pull_request?(payload['action'])
+      branch_name = get_branch_name(payload['pull_request']['head']['ref'])
+      sha = payload['pull_request']['head']['sha']
+      handle_pull_request(branch_name, sha)
     end
 
-    handle_pull_request if payload["pull_request"]
-    handle_repo_push_request if payload["ref"]
-    render :json => {:message => "Thanks!"}
+    if payload['ref']
+      branch_name = get_branch_name(payload['ref'])
+      sha = payload['after']
+      handle_repo_push_request(branch_name, sha)
+    end
   end
 
-  protected
+  def handle_github_request(payload)
+    repository = payload['repository']
+    pull_request = payload['pull_request']
+    namespace = repository['full_name'].split("/").first if repository['full_name'].present?
+    repo_info = {
+      host: 'git.github.com',
+      namespace: namespace,
+      name: repository['name'],
+      url: repository['html_url'] || repository['ssh_url']
```

Comment:
Since it should be guaranteed that the `ssh_url` will be present, I would just use that one and ignore the `html_url`.